src/osemgrep/cli_scan/Scan_subcommand.ml

```diff
@@ -142,6 +142,88 @@ let file_match_results_hook (conf : Scan_CLI.conf) (rules : Rule.rules)
           ~color_output:conf.force_color Format.std_formatter cli_matches)
       ~finally:(fun () -> Unix.lockf Unix.stdout Unix.F_ULOCK 0))
 
+(*****************************************************************************)
+(* Pretty Printing for CLI UX *)
+(*****************************************************************************)
+
+(* TODO: Update pysemgrep and osemgrep tests to match new output *)
+let new_cli_ux =
+  match !Env.v.user_agent_append with
+  | Some x -> (
+      match String.lowercase_ascii x with
+      | "pytest" -> false
+      | _ -> true)
+  | _ -> true
+
+let print_logo () : unit =
+  let logo =
+    Ocolor_format.asprintf
+      {|
+┌──── @{<green>○○○@} ────┐
+│ Semgrep CLI │
+└─────────────┘
+|}
+  in
+  Logs.app (fun m -> m "%s" logo);
+  ()
+
+let feature_status_str ~(enabled : bool) : string =
+  let status =
+    if enabled then Ocolor_format.asprintf {|@{<green>✔@}|}
+    else Ocolor_format.asprintf {|@{<red>✘@}|}
+  in
+  status
+
+let print_feature_section ~(includes_token : bool) ~(engine : Engine_type.t) :
+    unit =
+  let secrets_enabled =
+    match engine with
+    | PRO
+        Engine_type.
+          { secrets_config = Some Engine_type.{ allow_all_origins = _; _ }; _ }
+      ->
+        true
+    | OSS
+    | PRO Engine_type.{ secrets_config = None; _ } ->
+        false
+  in
+  let features =
+    [
+      ( ( "Semgrep OSS",
+          "Basic security coverage for first-party code vulnerabilities." ),
+        true );
+      ( ( "Semgrep Code (SAST)",
+          "Find and fix vulnerabilities in the code you write with advanced \
+           scanning and expert security rules." ),
+        includes_token );
+      ( ( "Semgrep Secrets",
+          "Detect and validate potential secrets in your code." ),
+        secrets_enabled );
+    ]
+  in
+  (* Print our set of features and whether each is enabled *)
+  List.iter
+    (fun tup ->
+      Logs.app (fun m ->
+          m "%s %s"
+            (snd tup |> feature_status_str)
+            (Ocolor_format.asprintf {|@{<bold>%s@}|} (fst (fst tup))));
+      Logs.app (fun m ->
+          m "  %s %s\n" (snd tup |> feature_status_str) (snd (fst tup))))
+    features;
+  ()
+
+let display_rule_source ~(rule_source : Rules_source.t) : unit =
+  let msg =
+    match rule_source with
+    | Configs _ ->
+        Ocolor_format.asprintf {|@{<bold>  %s@}|}
+          "Loading rules from registry..."
```

Comment:
Yep. I tried to place a generic enough message here without unpacking if the config was one of

```
| File of Fpath.t
| Dir of Fpath.t
| URL of Uri.t
| R of registry_config_kind
| A of app_config_kind
```

Do you have an alternative suggestion?

Reply:
Any reason _not_ to determine that here? We parse it a bit later but I don't really see a reason we couldn't parse it sooner so that `Configs` can have a `C.t list` instead of a `string list` (using `Rules_config.parse_config_string`; currently called in `rules_from_rules_source`) and then something like
```ocaml
| Configs xs when List.exists (function A _ | R _ -> true | _ -> false) xs -> (* current case *)
| Configs _ -> (* new case for custom local rules in files *)
```

Reply:
I don't remember why but having a string list and not parse the config string at parse-CLI-args time had some advantages.

Reply:
Maybe if we did it in cmdliner we get an error message less helpful than what we could give you in `Scan_subcommand.ml`? In any event I think by the start of `run_scan_conf` we could parse.